File: src/data/repositories/CategoriesRepository.ts
import RepositoryWithCrypt from './RepositoryWithCrypt';

import Category from "../models/Category";
import InvoiceRegistry from '../models/InvoiceRegistry';
import { Collections } from "../../data/firebase/Collections";

export default class CategoriesRepository extends RepositoryWithCrypt<Category> {

	constructor() {
		super(`${Collections.Users}/{userId}/${Collections.Categories}`, Category);
		this.addToCache(new Category(InvoiceRegistry.categoryId, "Fatura Cartão de Crédito", "creditcard", "#000000"));
	}

	public getAllRoots = async (): Promise<RootCategory[]> => {
		await this.waitInit();
		let rootCategories: RootCategory[] = [];
		let categories: Map<string, Category> = new Map();

		this.getCache().forEach((category) => {
			categories.set(category.id, category);
			if (!category.parentId) {
				rootCategories.push({
					...category,
					children: [],
				});
			}
		});

		rootCategories.forEach((root) => {
			categories.forEach((category) => {
				if (category.parentId === root.id) {
					category.color = root.color;
					category.icon = root.icon;
					root.children.push(category);
				}
			});
		});

		return rootCategories;
	};

	public override getLocalById(id?: string): Category | undefined {
		const category = super.getLocalById(id);
		if (!category) return undefined;
		if (!category?.icon && category?.parentId) {
			const root = this.getLocalById(category.parentId);
			if (!root) return category;
			category.color = root.color ?? "#000000";
			category.icon = root.icon ?? "folder";
		}
		return category;
	}
}

export interface RootCategory extends Category {
	children: Category[];
}

File: Android/app/src/main/kotlin/br/com/victorwads/goldenunicorn/data/repositories/CategoriesRepository.kt
package br.com.victorwads.goldenunicorn.data.repositories

import br.com.victorwads.goldenunicorn.data.firebase.Collections
import br.com.victorwads.goldenunicorn.data.models.Category
import br.com.victorwads.goldenunicorn.data.models.RootCategory
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.CollectionReference

class CategoriesRepository(
    userId: String? = FirebaseAuth.getInstance().currentUser?.uid
) : BaseRepository<Category>(Category::class.java) {

    override val cacheDuration: Long = 0
    override val ref: CollectionReference

    init {
        val finalUserId = userId ?: throw Error("Invalid userId")
        ref = bd
            .collection(Collections.Users)
            .document(finalUserId)
            .collection(Collections.Categories)
    }

    suspend fun getAll(forceCache: Boolean = false): List<Category> =
        getAllItems(forceCache)

    fun getAllRoots(): List<RootCategory> {
        val categories = cache.values
        val roots = mutableMapOf<String, RootCategory>()

        categories.filter { it.parentId == null }.forEach { cat ->
            roots[cat.id] = RootCategory(
                id = cat.id,
                name = cat.name,
                icon = cat.icon,
                color = cat.color,
            )
        }

        categories.filter { it.parentId != null }.forEach { cat ->
            val parent = cat.parentId?.let { roots[it] } ?: return@forEach
            parent.children.add(cat.copy(color = parent.color, icon = parent.icon))
        }
        return roots.values.toList()
    }
}
